modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java

```diff
@@ -297,25 +299,50 @@ private void addFeatureInternal(SimpleFeature feature) {
     protected List<Name> createTypeNames() {
         synchronized (memory) {
             List<Name> typeNames = new ArrayList<Name>();
-            
+
             for (Iterator<String> i = schema.keySet().iterator(); i.hasNext();) {
                 typeNames.add( new NameImpl(namespaceURI, i.next()));
             }
             return typeNames;
         }
     }
-    
+
     protected ContentFeatureSource createFeatureSource(ContentEntry entry) {
         return createFeatureSource(entry, Query.ALL);
     }
-    
+
     protected ContentFeatureSource createFeatureSource(ContentEntry entry, Query query) {
-        return new MemoryFeatureStore(entry, query);
+        return new MemoryFeatureStore(entry, query) {
+            @Override protected QueryCapabilities buildQueryCapabilities() {
+                return new QueryCapabilities() {
+                    @Override public boolean isUseProvidedFIDSupported() {
+                        return true;
+                    }
+                };
+            }
+
+            //need to actually use the provided FID
+            @Override protected FeatureWriter<SimpleFeatureType, SimpleFeature> getWriterInternal(
+                Query query, int flags) throws IOException {
+                return new MemoryFeatureWriter(getState(), query) {
+                    @Override public void write() throws IOException {
```

Comment:
Same here, please fix MemoryFeatureWriter directly, no need for anonymous subclass.